Source/Core/VideoCommon/Assets/DirectFilesystemAssetLibrary.h

```diff
@@ -0,0 +1,37 @@
+// Copyright 2023 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+#include <chrono>
+#include <filesystem>
+#include <map>
+#include <string>
+
+#include "VideoCommon/Assets/CustomAssetLibrary.h"
+
+namespace VideoCommon
+{
+class CustomTextureData;
+
+// This class implements 'CustomAssetLibrary' and loads any assets
+// directly from the filesystem
+class DirectFilesystemAssetLibrary final : public CustomAssetLibrary
+{
+public:
+  LoadInfo LoadTexture(const AssetID& asset_id, CustomTextureData* data) override;
+
+  // Gets the latest time from amongst all the files in the asset map
+  TimeType GetLastAssetWriteTime(const AssetID& asset_id) const override;
+
+  // Assigns the asset id to a map of files, how this map is read is dependent on the data
+  // For instance, a raw texture would expect the map to have a single entry and load that
+  // file as the asset.  But a model file data might have its data spread across multiple files
+  void SetAssetIDMapData(const AssetID& asset_id,
+                         std::map<std::string, std::filesystem::path> asset_path_map);
```

Comment:
Okay, I realize this has now resulted from my note about the asset_id being a file system path, but this is... pretty strange to me. Does that mean you now have to 'prime' every asset you load before you load it by calling this function? How would this work in a situation where you don't know (or care) what library is hiding behind your `CustomAssetLibrary` pointer you have?

In fact, okay, maybe step back. Can you give me a little example (pseudocode is fine) on how exactly the general flow here is supposed to work?

Reply:
> Does that mean you now have to 'prime' every asset you load before you load it by calling this function?

Yes.

Examples.

**Graphics mods**:

Plan is to have a blob in the json called "assets" with the mapping from name (assetID) to path.  Eventually an editor would probably define the name, maybe as a GUID (I think this is what Unity does).

**Texture packs**:

You'd iterate the filesystem on load, looking for the files.  Then have the textureid as the assetID and map that to the path on the disk.

-----------

A really dumb mapper might be like:

```
for (const auto& file_path : hires_texture_paths)
{
  std::map<std::string, std::filesystem::path> asset_mapper;
  // The map usage is if we had multiple files we wanted to map to one asset
  // we only have one file, so just make a dummy key, it doesn't matter in the texture loading code
  asset_mapper[""] = file_path;
  m_directfilesystem_library->SetAssetIDMapData(file_path.name(), std::move(asset_mapper);
}
```

I'll admit it is a bit klunky but after your comment, I realized a filesystem path doesn't make a lot of sense either.  There's times we don't know the path!

This might be something we have to grow into after we get all the cases laid out (I haven't finished designing material/shader assets yet).  I am open to ideas.